Support SSL Certificate Hot Reloading

elcallio opened the pull request:
Fixes #6067

Requires seastar patch set "Reloadable TLS certificates"

Makes the scylla endpoint initializations that support TLS use reloadable certificate stores, watching used cert + key files for changes, and reload iff modified. 

Tests in separate dtest set.

avikivity commented on message/messaging_service.hh:
```diff
@@ -254,6 +255,7 @@ public:
             memory_config mcfg, scheduling_config scfg, bool sltba = false, bool listen_now = true);
     ~messaging_service();
 public:
+    future<> start_listen_ex();
```
Can we avoid having two entry points? e.g. rename start_listen() to something private and internal?

avikivity commented on message/messaging_service.cc:
```diff
@@ -379,7 +399,9 @@ messaging_service::messaging_service(gms::inet_address ip
+    if (listen_now && credentials) {
+        throw std::invalid_argument("Cannot start from contructor when TLS is enabled");
+    }
     if (listen_now) {
```
Let's drop listen_now() to avoid having two code paths.

avikivity commented on alternator/server.cc:
```diff
@@ -414,7 +414,13 @@ future<> server::init(net::inet_address addr, std::optional<uint16_t> port, std:
+                    if (ep) {
+                        slogger.warn("Exception loading {}: {}", files, ep);
+                    } else {
+                        slogger.info("Reloaded {}", files);
```
You can use info0() instead to avoid having 98 repetitions of the message.

elcallio replied:
Yes, but that would generate potential confusion with the warning on exception. What can happen is a user only replaces the PEM and not the KEY file of a cert pair, than you will get an exception "mismatched certs". Then he completes his file mods and replaces KEY as well, so reload succeeds, and we print the joyful message. 
But if this was in a race condition where some shard is lagging, we would get could get a shard0 "success", but a shardX fail (followed by a silence success). Me thinks confusing to log reader...

avikivity commented:
Please rebase on top of master so we pick up the seastar changes.

elcallio commented:
Updated: 

- Rebased on newest master
- Made "messaging_service::start_listen" non-optional (remove "listen_now" option)

Did not change the logging, because of reason above. Could optionally create shared wrapper type that keeps state of whether it was called with errors and outputs info even on non-zero shard to make logging consistent for user. Maybe follow-up?

avikivity commented:
Queued.
